### PR TITLE
Implements import API, list import, and site import

### PR DESCRIPTION
**Pull request by cperryk**
## programmatic import api

* Creates a programmatic import API for importing any asset (cmpt, page, list, user, or uri) or class of assets from one site to another as well as data from files or streams. Extracts import functions from `cmd/import`, which is now concerned only with translating user commands into import API calls and logging the results of those commands.
* Implements site-to-site import. This fetches pages via the pages index rather than the /pages endpoint. It supports a `published` option to import published pages.
* Implements an `overwriteLayouts` option for single-page and site-to-site imports. These functions now by default do not overwrite layouts or layout descendants. Setting `overwriteLayouts` overwrites them.

## import cmd

* Refactors cmd/import logging to render dots on a single refreshing line (else site-to-site import prints thousands of dots).
* Removes `--page` and `--component` options  and adds a `--url` (or `-u`) option, which sets the command to import the specified URL, whether it is a page, cmpt, list, user, or uri. (I've also changed this for the `export` cmd). 

## utils/rest changes

* `rest.get` no longer swallows errors. This makes it consistent with rest.put and to allows dependents to decide error flow.
* added `rest.post` so we can query _search Elastic endpoint for pages (for site-to-site import)
* `rest.put` now converts objects to JSON automatically for conciseness
* Renames `result` property of rest result objects to `status` to distinguish it from the result object itself (e.g. `result.result` is confusing

## input-clay changes
 
* Added `streamPageUris`, `streamListUris`, `streamUserUris`, and `streamUris`
* Renamed some functions to distinguish them from the functions of the import API. (e.g. `importUrl` is now `streamAssets`).
* `streamAssets` now supports all Clay asset types (cmpt, page, list, user, uri)
* `getDeepPage` has been renamed `streamPageAssets` and does _not_ GET page-level cmpts; this is so URLs are deduped within `streamAssets` to avoid unecessary GETs.

## What's next

* Implement an option in site-to-site import to cherrypick asset types
* Create an `export` API and refactor the import API to use it (perhaps replacing input-clay)

### CODE REVIEW

File: test/setup.js [hunk 1]
```diff
@@ -1,4 +1,3 @@
-'use strict';
```
Comment (amycheng) [inline test/setup.js]: why was this removed?

File: package.json [hunk 2]
```diff
@@ -44,8 +44,9 @@
     "bluebird": "^3.5.0",
     "byline": "^5.0.0",
     "chalk": "^1.1.3",
-    "clay-utils": "^1.1.1",
+    "clay-utils": "^1.3.0",
     "console-log-level": "^1.4.0",
+    "deep-equal": "^1.0.1",
```
Comment (amycheng) [inline package.json]: any reason why you're using this instead of `lodash/_.isEqual`?

**Comment by amycheng** [comment]
# Acceptance Criteria
aka stuff Amy will be QA-ing during this code review

- [ ] site A to site B import
- [ ] site A to site A import
- [ ] page import still works
- [ ] component import 
- [ ] layouts are not overwritten during page import
- [ ] layouts can be overwritten during page import
- [ ] works with Webpack

File: package.json (hunk 2 above)
Comment (cperryk) [reply package.json]: Good idea, will change.

File: test/setup.js (hunk 1 above)
Comment (cperryk) [reply test/setup.js]: breaks linting

**Comment by amycheng** [comment]
Bugs with -url single page import
=========================

using this command: `clay import -u [url] http://localhost.thecut.com:3001 -k [qa access key] --verbose`

https://www.thecut.com/pages/cj95flr2500b0jwye7x0jgyo2 
`500 partial undefined` (I've seen this bug pop up for some other pages too)

https://www.thecut.com/pages/cj94fh8fj00cizgy6bmghjwps@published
`article instance not found` (also saw this bug for other articles)

`https://www.thecut.com/pages/cj8xd4z0k0051c5y6epb67zqm@published`
while this was a successful import, a slug is not being generated and doesn't show up in `/news`. Also when I try to republish the page, Kiln throws an error

**Comment by amycheng** [comment]
I'm also getting errors when importing articles like `nymag.com/strategist/pages/cj95u30oe00azu7y6smtmsss6@published`.

When importing `greatest-hits` or other components that rely on the destination's ES index, the importer will fail because the destination's ES index does not match the source's ES index. This is what I suspect is happening.

So, I wonder if there's a way to gracefully handle these kind of errors and continue on with the import?

**Comment by amycheng** [comment]
`http://nymag.com/daily/intelligencer/pages/cj95wiws300itk9yevo4oi95y` is another page whose `@published` instance is imported incorrectly. When importing the `@published` instance, a slug isn't being generated and when re-publishing, Kiln throws an error. However, importing the draft instance and manually publishing works.

**Comment by amycheng** [comment]
Bugs with site-to-site import
=====================

Site-to-Site import bugs

when running:
`clay import -s http://localhost.dev.nymag.biz:3001/daily/intelligencer -l 10 http://localhost.thecut.com:3001 -k [QA Key] -sourceKey [QA Key]`

I get the folowing error:
```
/Users/amycheng/Desktop/qa/claycli/node_modules/yargs/yargs.js:1079
      else throw err
           ^

TypeError: url.replace is not a function
    at normalizeSite (/Users/amycheng/Desktop/qa/claycli/lib/utils/config.js:79:25)
    at Object.getSite (/Users/amycheng/Desktop/qa/claycli/lib/utils/config.js:98:39)
    at normalizeArgv (/Users/amycheng/Desktop/qa/claycli/lib/cmd/import.js:217:24)
    at Object.handler (/Users/amycheng/Desktop/qa/claycli/lib/cmd/import.js:245:3)
    at Object.self.runCommand (/Users/amycheng/Desktop/qa/claycli/node_modules/yargs/lib/command.js:233:22)
    at Object.Yargs.self._parseArgs (/Users/amycheng/Desktop/qa/claycli/node_modules/yargs/yargs.js:990:30)
    at Object.get [as argv] (/Users/amycheng/Desktop/qa/claycli/node_modules/yargs/yargs.js:927:19)
    at Object.<anonymous> (/Users/amycheng/Desktop/qa/claycli/index.js:42:5)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:389:7)
    at startup (bootstrap_node.js:149:9)
    
```

**Comment by cperryk** [comment]
@amycheng it's `--sourceKey` not `-sourceKey`. Needs two hyphens

**Comment by amycheng** [comment]
@cperryk doh

**Comment by amycheng** [comment]
@cperryk is there a reason why `soureKey` needs two dashes? the other flags just use one dash. It was very easy for me to make that typo because I assumed `sourceKey` used one day like the other flags.

**Comment by amycheng** [comment]
@cperryk site-to-site import ran without errors but published articles from DI, were NOT published when they were imported to the Cut. Is this a feature or a bug?

This is how my local environment was set-up:
- use `scratch-cli` to import first-runs into a local instance of the Cut and DI
- use `clay-cli` to import a few articles into both 
- use `clay-cli` to do an import of DI into the Cut
